internal/service/serverlessinstance/data_source_serverless_instances.go

```diff
@@ -25,91 +24,87 @@ func PluralDataSource() *schema.Resource {
 				Computed: true,
 				Elem: &schema.Resource{
-					Schema: returnServerlessInstanceDSSchema(),
+					Schema: dataSourceSchema(),
 				},
 			},
 		},
 	}
 }
 
-func dataSourceMongoDBAtlasServerlessInstancesRead(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+func dataSourcePluralRead(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	connV2 := meta.(*config.MongoDBClient).AtlasV2
 	projectID, projectIDOK := d.GetOk("project_id")
-
 	if !(projectIDOK) {
 		return diag.Errorf("project_id must be configured")
 	}
-
-	options := &matlas.ListOptions{
-		ItemsPerPage: 500,
-		IncludeCount: true,
+	options := &admin.ListServerlessInstancesApiParams{
+		ItemsPerPage: conversion.IntPtr(500),
+		IncludeCount: conversion.Pointer(true),
+		GroupId:      projectID.(string),
 	}
 
-	serverlessInstances, err := getServerlessList(ctx, meta, projectID.(string), options, 0)
+	serverlessInstances, err := getServerlessList(ctx, connV2, options, 0)
 	if err != nil {
 		return diag.Errorf("error getting serverless instances information: %s", err)
 	}
 
 	flatServerlessInstances := flattenServerlessInstances(serverlessInstances)
-
 	if err := d.Set("results", flatServerlessInstances); err != nil {
 		return diag.Errorf("error setting `results` for serverless instances: %s", err)
 	}
 
 	d.SetId(id.UniqueId())
-
 	return nil
 }
 
-func getServerlessList(ctx context.Context, meta any, projectID string, options *matlas.ListOptions, obtainedItemsCount int) ([]*matlas.Cluster, error) {
-	// Get client connection.
-	var list []*matlas.Cluster
-	options.PageNum++
-	conn := meta.(*config.MongoDBClient).Atlas
-
-	serverlessInstances, _, err := conn.ServerlessInstances.List(ctx, projectID, options)
+func getServerlessList(ctx context.Context, connV2 *admin.APIClient, options *admin.ListServerlessInstancesApiParams, obtainedItemsCount int) ([]admin.ServerlessInstanceDescription, error) {
+	if options.PageNum == nil {
+		options.PageNum = conversion.IntPtr(1)
+	} else {
+		*options.PageNum++
+	}
+	var list []admin.ServerlessInstanceDescription
+	serverlessInstances, _, err := connV2.ServerlessInstancesApi.ListServerlessInstancesWithParams(ctx, options).Execute()
 	if err != nil {
 		return list, fmt.Errorf("error getting serverless instances information: %s", err)
 	}
 
-	list = append(list, serverlessInstances.Results...)
-	obtainedItemsCount += len(serverlessInstances.Results)
+	list = append(list, serverlessInstances.GetResults()...)
+	obtainedItemsCount += len(serverlessInstances.GetResults())
 
-	if serverlessInstances.TotalCount > options.ItemsPerPage && obtainedItemsCount < serverlessInstances.TotalCount {
-		instances, err := getServerlessList(ctx, meta, projectID, options, obtainedItemsCount)
+	if serverlessInstances.GetTotalCount() > *options.ItemsPerPage && obtainedItemsCount < *serverlessInstances.TotalCount {
+		instances, err := getServerlessList(ctx, connV2, options, obtainedItemsCount)
```

Comment:
in my opinion using recursion here is unnecessary and makes the code more confusing and not straight forward. I would change it to a for loop [equivalent to a "do while"](https://stackoverflow.com/questions/32834661/how-to-replicate-do-while-in-go/32844744#32844744) (so we always do the first iteration of the loop)

Reply:
to be honest I just "mimicked" the current code using connv2,  as we'll have a coming project to improve plural data sources. so i'd prefer to leave this optimization for a future PR